taichi/transforms/alg_simp.cpp

```diff
@@ -327,6 +328,22 @@ class AlgSimp : public BasicStmtVisitor {
   }
 };
 
+class AlgSimpPass : public Pass {
+ public:
+  static const PassID id;
+  class Result : public Pass::Result {
+   public:
+    bool modified;
+    explicit Result(bool modified) : modified(modified) {
+    }
+  };
+  static void run(const PassContext &ctx, IRNode *module, PassManager *mgr) {
+    auto result = AlgSimp::run(module, ctx.get_config().fast_math);
+    mgr->put_pass_result<AlgSimpPass>(std::make_unique<Result>(result));
+  }
+};
+const PassID AlgSimpPass::id = "AlgSimpPass";
```

Comment:
I don't like this `PassID` design since it needs writing two separate lines for each pass. What about using `std::typeindex(typeid(XPass))` directly?

Reply:
> Could you use `static constexpr char *` as the ID type? This should allow you to write within the class?

`std::string` doesn't have a constexpr constructor, so I can't write it within the class.

> Pass ID may get used elsewhere. When that grows, this could cause us to write `std::type_index(...)` everywhere. (IIUC, your point is to avoid having an `ID` member? In general it is necessary to have an identifier for each pass) It's also less searchable than writing out the ID explicitly.

My point is to minimize the repeated part of code (or, similar part with only the class name different) that each pass (class) shares. I thought there may be an easier way to implement an id for each pass than this:
```C++
static const PassID id;
...
const PassID AlgSimpPass::id = "AlgSimpPass";
```

Reply:
To use `std::type_index`, I mean something like this:
```C++
class Pass {
  ...
  std::type_index get_pass_id() const {
    return std::type_index(typeid(*this));
  }
};
```

Reply:
I see. I think this is still a class property than an instance one. Pass ID might get used elsewhere as an ID. For example, we might have pass dependencies that are specified through the names/IDs. Using `type_index` gives no information of what the pass name will be.

Reply:
Right, I agree that we may need the pass name somewhere. If there are no convenient macros, maybe we can only repeat the pass name in the pass ID.